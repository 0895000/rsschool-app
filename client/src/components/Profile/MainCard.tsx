import * as React from 'react';
import { GithubAvatar } from 'components';
import {
  Card,
  Typography,
} from 'antd';
import { GeneralInfo, ConfigurableProfilePermissions } from '../../../../common/models/profile';
import VisibilitySettingsDrawer from './VisibilitySettingsDrawer';
import { ChangedSettings } from 'pages/profile';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';

const { Title, Paragraph } = Typography;

import {
  GithubFilled,
  EnvironmentFilled,
  EditOutlined,
  SettingOutlined,
} from '@ant-design/icons';

type Props = {
  data: GeneralInfo;
  isEditingModeEnabled: boolean;
  permissionsSettings?: ConfigurableProfilePermissions;
  onSettingsChange: (event: CheckboxChangeEvent, changedSettings: ChangedSettings) => void;
};

type State = {
  isSettingsVisible: boolean;
}

class MainCard extends React.Component<Props, State> {
  state = {
    isSettingsVisible: false,
  }

  private showSettings = () => {
    this.setState({ isSettingsVisible: true });
  }

  private hideSettings = () => {
    this.setState({ isSettingsVisible: false });
  }

  private filterPermissions = ({ isProfileVisible }: Partial<ConfigurableProfilePermissions>) => ({
    isProfileVisible,
  })

  render() {
    const { isEditingModeEnabled, permissionsSettings, onSettingsChange } = this.props;
    const { githubId, name, locationName } = this.props.data;
    const { isSettingsVisible } = this.state;

    return (
      <>
        <Card
          actions={isEditingModeEnabled ? [
            <EditOutlined key="main-card-actions-edit"/>,
            <SettingOutlined key="main-card-actions-settings" onClick={this.showSettings} />,
          ] : undefined}
        >
          <GithubAvatar size={96} githubId={githubId} style={{ margin: '0 auto 10px', display: 'block' }} />
          <Title level={1} style={{ fontSize: 24, textAlign: 'center', margin: 0 }}>{name}</Title>
          <Paragraph style={{ textAlign: 'center', marginBottom: 20 }}>
            <a target="_blank" href={`https://github.com/${githubId}`} style={{ marginLeft: '-14px', fontSize: 16 }}>
              <GithubFilled /> {githubId}
            </a>
          </Paragraph>
          <Paragraph style={{ textAlign: 'center', margin: 0 }}>
            <span style={{ marginLeft: '-14px' }}>
              <EnvironmentFilled /> {locationName}
            </span>
          </Paragraph>
          {
            isEditingModeEnabled &&
              <VisibilitySettingsDrawer
                isSettingsVisible={isSettingsVisible}
                hideSettings={this.hideSettings}
                permissionsSettings={permissionsSettings ? this.filterPermissions(permissionsSettings) : undefined}
                onSettingsChange={onSettingsChange}
              />
          }
        </Card>
      </>
    );
  }
}

export default MainCard;
