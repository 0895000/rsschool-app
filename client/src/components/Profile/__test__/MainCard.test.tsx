import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import MainCard from '../MainCard';

describe('MainCard', () => {
  describe('Should render correctly', () => {
    it('if is editing mode disabled', () => {
      const output = shallow(
        <MainCard
          data={{
            name: 'Petr Pervyi',
            githubId: 'piter',
            locationName: 'SPB',
            locationId: '1',
          }}
          isEditingModeEnabled={false}
          onPermissionsSettingsChange={() => {}}
          onProfileSettingsChange={() => {}}
        />,
      );
      expect(shallowToJson(output)).toMatchSnapshot();
    });
  });

  // const wrapper = shallow(
  //   <MainCard
  //     data={{
  //       name: 'Petr Pervyi',
  //       githubId: 'piter',
  //       locationName: 'SPB',
  //       locationId: '1',
  //     }}
  //     isEditingModeEnabled={false}
  //     onPermissionsSettingsChange={() => {}}
  //     onProfileSettingsChange={() => {}}
  //   />);
  // const instance = wrapper.instance();
  // describe('showVisibilitySettings', () => {
  //   it('should set "state.isVisibilitySettingsVisible" as "true"', () => {
  //     expect(instance.state.isVisibilitySettingsVisible).toBe(false);
  //     instance.showVisibilitySettings();
  //     expect(instance.state.isVisibilitySettingsVisible).toBe(true);
  //   });
  // });
  // describe('hideVisibilitySettings', () => {
  //   it('should set "state.isVisibilitySettingsVisible" as "false"', () => {
  //     instance.state.isVisibilitySettingsVisible = true;
  //     expect(instance.state.isVisibilitySettingsVisible).toBe(true);
  //     instance.hideVisibilitySettings();
  //     expect(instance.state.isVisibilitySettingsVisible).toBe(false);
  //   });
  // });
  // describe('showProfileSettings', () => {
  //   it('should set "state.isProfileSettingsVisible" as "true"', () => {
  //     expect(instance.state.isProfileSettingsVisible).toBe(false);
  //     instance.showProfileSettings();
  //     expect(instance.state.isProfileSettingsVisible).toBe(true);
  //   });
  // });
  // describe('hideProfileSettings', () => {
  //   it('should set "state.isProfileSettingsVisible" as "false"', () => {
  //     instance.state.isProfileSettingsVisible = true;
  //     expect(instance.state.isProfileSettingsVisible).toBe(true);
  //     instance.hideProfileSettings();
  //     expect(instance.state.isProfileSettingsVisible).toBe(false);
  //   });
  // });
});
