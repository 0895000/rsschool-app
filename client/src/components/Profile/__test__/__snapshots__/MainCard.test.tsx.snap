// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainCard Should render correctly if is editing mode disabled 1`] = `
<Fragment>
  <Card>
    <GithubAvatar
      githubId="piter"
      size={96}
      style={
        Object {
          "display": "block",
          "margin": "0 auto 10px",
        }
      }
    />
    <Title
      level={1}
      style={
        Object {
          "fontSize": 24,
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      Petr Pervyi
    </Title>
    <Paragraph
      style={
        Object {
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      <a
        href="https://github.com/piter"
        style={
          Object {
            "fontSize": 16,
            "marginLeft": "-14px",
          }
        }
        target="_blank"
      >
        <ForwardRef(GithubFilled) />
         
        piter
      </a>
    </Paragraph>
    <Paragraph
      style={
        Object {
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      <span
        style={
          Object {
            "marginLeft": "-14px",
          }
        }
      >
        <ForwardRef(EnvironmentFilled) />
         
        Minsk, Belarus
      </span>
    </Paragraph>
  </Card>
</Fragment>
`;

exports[`MainCard Should render correctly if is editing mode enabled 1`] = `
<Fragment>
  <Card
    actions={
      Array [
        <ForwardRef(EditOutlined)
          onClick={[Function]}
        />,
        <ForwardRef(SettingOutlined)
          onClick={[Function]}
        />,
      ]
    }
  >
    <GithubAvatar
      githubId="piter"
      size={96}
      style={
        Object {
          "display": "block",
          "margin": "0 auto 10px",
        }
      }
    />
    <Title
      level={1}
      style={
        Object {
          "fontSize": 24,
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      Petr Pervyi
    </Title>
    <Paragraph
      style={
        Object {
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      <a
        href="https://github.com/piter"
        style={
          Object {
            "fontSize": 16,
            "marginLeft": "-14px",
          }
        }
        target="_blank"
      >
        <ForwardRef(GithubFilled) />
         
        piter
      </a>
    </Paragraph>
    <Paragraph
      style={
        Object {
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      <span
        style={
          Object {
            "marginLeft": "-14px",
          }
        }
      >
        <ForwardRef(EnvironmentFilled) />
         
        Minsk, Belarus
      </span>
    </Paragraph>
    <PermissionsSettingsDrawer
      hideSettings={[Function]}
      isSettingsVisible={false}
      onPermissionsSettingsChange={[MockFunction]}
    />
    <ProfileSettingsDrawer
      content={
        <div>
          <p
            style={
              Object {
                "fontSize": 18,
                "marginBottom": 5,
              }
            }
          >
            <Text
              strong={true}
            >
              Name:
            </Text>
          </p>
          <p
            style={
              Object {
                "marginBottom": 20,
              }
            }
          >
            <Input
              onChange={[Function]}
              placeholder="Firstname Lastname"
              type="text"
              value="Petr Pervyi"
            />
          </p>
          <p
            style={
              Object {
                "fontSize": 18,
                "marginBottom": 5,
              }
            }
          >
            <Text
              strong={true}
            >
              Location:
            </Text>
          </p>
          <div
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            <LocationSelect
              location={
                Object {
                  "cityName": "Minsk",
                  "countryName": "Belarus",
                }
              }
              onChange={[Function]}
            />
          </div>
        </div>
      }
      hideSettings={[Function]}
      isSettingsVisible={false}
    />
  </Card>
</Fragment>
`;
